File: swift/708_InsertIntoASortedCircularLinkedList.swift
/**
 * Definition for a Node.
 * public class Node {
 *     public var val: Int
 *     public var next: Node?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func insert(_ head: Node?, _ insertVal: Int) -> Node? {
        if head == nil {
            var newHead: Node? = Node(insertVal)
            newHead?.next = newHead
            return newHead
        }
        
        func insertNode(_ currentNode: Node?, _ nextNode: Node?) -> Node? {
            var curr = currentNode, next = nextNode
            var newNode: Node? = Node(insertVal)
            curr?.next = newNode
            newNode?.next = next
            return head
        }

        var curr = head
        while curr != nil {
            var currVal = curr?.val as! Int
            var nextVal = curr?.next?.val as! Int
            if currVal == nextVal {
                var start: Node? = curr
                while curr?.next !== start && currVal == nextVal {
                    curr = curr?.next
                    currVal = curr?.val as! Int
                    nextVal = curr?.next?.val as! Int
                }
                if currVal == nextVal || insertVal == currVal {
                    return insertNode(curr, curr?.next)
                }
            }
            if currVal < nextVal {
                if insertVal >= currVal && insertVal <= nextVal {
                    return insertNode(curr, curr?.next)
                }
            } else if currVal > nextVal {
                if insertVal >= currVal || insertVal <= nextVal {
                    return insertNode(curr, curr?.next)
                }
            }
            curr = curr?.next
        }
        return head
    }
}

File: java/708_InsertIntoASortedCircularLinkedList.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node next;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _next) {
        val = _val;
        next = _next;
    }
};
*/

class Solution {
    public Node insert(Node head, int insertVal) {
        if (head == null) {
            Node node = new Node(insertVal);
            node.next = node;
            return node;
        }
        Node curr = head;
        while (curr != null) {
            int currVal = curr.val, nextVal = curr.next.val;
            if (currVal == nextVal) {
                if (insertVal == currVal) {
                    insertNext(curr, curr.next, insertVal);
                    return head;
                }
                Node start = curr;
                while (currVal == nextVal && start != curr.next) {
                    curr = curr.next;
                    currVal = curr.val;
                    nextVal = curr.next.val;
                }
                if (curr.next == start) {
                    insertNext(curr, curr.next, insertVal);
                    return head;
                }
            }
            
            if (currVal < nextVal) {
                if (currVal <= insertVal && insertVal <= nextVal) {
                    insertNext(curr, curr.next, insertVal);
                    return head;
                }
            } else if (currVal > nextVal) {
                if (currVal <= insertVal || nextVal >= insertVal) {
                    insertNext(curr, curr.next, insertVal);
                    return head;
                }
            }
            curr = curr.next;
        }
        return head;
    }
    
    public Node insertNext(Node curr, Node next, int insertVal) {
        Node node = new Node(insertVal);
        node.next = next;
        curr.next = node;
        return node;
    }
}
